README
LaTeX2HTML README
=================

Contents
********

Overview
Pointers to the User Manual
Requirements
Installation
Troubleshooting
Support and More Information

Overview
********

The LaTeX2HTML translator: 

 o breaks up a document into one or more components as specified by
   the user, 
 o provides optional iconic navigation panels on every page which
   contain links to other parts of the document,  
 o handles inlined equations, right-justified
   numbered equations, tables, or figures and any arbitrary environment, 
 o can produce output suitable for browsers that support inlined images
   or character based browsers (as specified by the user), 
 o handles definitions of new commands, environments, and theorems
   even when these are defined in external style files, 
 o handles footnotes, tables of contents, lists of figures and tables,
   bibliographies, and can generate an Index, 
 o translates cross-references into hyperlinks and extends the
   LaTeX cross-referencing mechanism to work not just
   within a document but between documents which may reside in
   remote locations, 
 o translates accent and special character
   commands to the equivalent ISO-LATIN-1
   character set where possible, 
 o recognizes hypertext links (to multimedia resources or arbitrary
   internet services such as sound/video/ftp/http/news) and links which
   invoke arbitrary program scripts, all expressed as LaTeX commands, 
 o recognizes conditional text which is intended only for the hypertext
   version, or only for the paper (DVI) version, 
 o can include raw HTML in a LaTeX document (e.g. in order to specify
   interactive forms), 
 o can deal sensibly with all the commands and environments commonly used 
   with LaTeX as summarized at the back of the LaTeX blue book [1],
   and many of the packages described in the LaTeX Companion, and others. 
 o will try to translate any document with embedded LaTeX commands
   irrespective of whether it is complete or syntactically legal. 

Pointers to the User Manual
***************************

The LaTeX2HTML program includes its own manual page. 
The manual page can be viewed by saying "perldoc latex2html"
or "latex2html -help".

See the online documentation at 
   http://www-texdev.ics.mq.edu.au/l2h/docs/manual/
for more information and examples.

Other useful links can be found at:  www.latex2html.org
and at the mailing-list site:
	http://tug.org/mailman/listinfo/latex2html

In particular see the pages:
 support.html , Snode1.html , Snode2.html , Snode3.html 
for instructions on how to install the program 
and make your own local copy of the manual in HTML.

Requirements
************

Please consult the section "Requirements" of the online manual at

   http://www-texdev.ics.mq.edu.au/l2h/docs/manual/Snode2.html

for more information as well as *active* links to any utilities
that you may require. 


The requirements for using LaTeX2HTML depend on the kind of
translation it is asked to perform as follows: 

 1. LaTeX commands but without equations, figures, tables, etc. 
    o Perl 5.003 (Perl5 Patch level 3) or higher.
------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    o DBM or NDBM, the Unix DataBase Management system.
      Alternatively, Perl5's SDBM DataBase system.
------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      Do not care unless you get misconfiguration errors from LaTeX2HTML.

 2. LaTeX commands with equations, figures, tables, etc. 
   As above plus 
    o latex (version 2e recommended but 2.09 acceptable), 
    o dvips (version 5.516 or later) or dvipsk.
      Version 5.62 or higher enhances the performance of image creation
      with a *significant* speed-up. See l2conf.pm for this
      after you are done with the installation.
      Do not use the 'dvips -E' feature unless you have 5.62, else you
      will get broken images.
------^^^^
    o gs (Ghostscript version 4.03 or later),
------------------------------^^^^
      with the ppmraw device driver, or even better pnmraw.
      Upgrade to 5.10 or later if you want to go sure about seldom problems
      with 4.03 to avoid (yet unclarified).
    o The netpbm library (ftp://ftp.x.org/R5contrib/).
      Netpbm 1 March 1994 is recommended. Check with 'pnmcrop -version'.
      Some of the filters in those libraries are used during the postscript
      to image conversion.
    o If you want PNG images, you need pnmtopng (current version is 2.31).
      It is not part of netpbm and requires libpng (version 0.89c) and 
      libz (1.0.4). pnmtopng supports transparency and interlace mode.
      Hurray!!! Netscape 4.04 has been reported to grok PNG images!
------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
      That means your PNG option is not longer ahead of its time!
      Unfortunately Netscape still does not make use of the alpha
      channel... still not transparency.

 3. Transparent inlined GIFs
   If you dislike the gray background color of the generated inlined images
   then the best thing you can do is get the netpbm library (instead of
   the older pbmplus) OR install the giftrans filter by Andreas Ley
   <ley@rz.uni-karlsruhe.de>. Version 1.10.2 is known to work without
   problems but later versions should also be OK.

   LaTeX2HTML also supports the shareware program giftool (by Home Pages, Inc.,
   version 1.0), too. It can also create interlaced GIFs.

Because by default the translator makes use of inlined images in the final
HTML output, it would be better to have a viewer which supports the <IMG>
tag, such as NCSA Mosaic. If only a character based browser is available or
if you want the generated documents to be more portable then the translator
can be used with the -ascii_mode option. 

If ghostscript or netpbm library are not available
it is still possible to use the translator with the -no_images option. 

If you intend to use any of the special features of the translator 
then you have to include the html.sty file in any LaTeX documents that
use them. 


Installation
************

Please consult the section "Installing LaTeX2HTML" of the online manual at

    http://www-texdev.ics.mq.edu.au/l2h/docs/manual/Snode3.html

Also consult the page at

   http://tug.org/mailman/listinfo/latex2html

for links to special instructions on some platforms.


To install LaTeX2HTML please read the file INSTALL.

Troubleshooting
***************

Please refer to the FAQ file that came with your distribution.


Support and More Information
****************************

A LaTeX2HTML mailing list has been set up at the TeX User Group (TUG). 
To join the list, visit the web-page at:

   http://tug.org/mailman/listinfo/latex2html

and follow the instructions found there.

If this is not possible for some reason, then send a message to: 
            latex2html-request@tug.org 
with the contents 
            subscribe 


To be removed from the list follow the instructions at:

   http://tug.org/mailman/listinfo/latex2html

If this is not possible for some reason, then send a message to: 
           latex2html-request@tug.org 
with the contents 
           unsubscribe


An archive of the mailing list, from 1999 onwards,
can be browsed at:

   http://tug.org/pipermail/latex2html/

This mailing list also has a searchable online archive,
from 1994 up until 2003-08, at

   http://www.xray.mpe.mpg.de/mailing-lists/latex2html/


Enjoy!


Original Author:
  Nikos Drakos <nikos@cbl.leeds.ac.uk>
  Computer Based Learning Unit
  University of Leeds.

Most Recent Author:
  Ross Moore <ross@maths.mq.edu.au>
  Mathematics Department
  Macquarie University, Sydney.

Former Authors:
  Marek Rouchal <marek@saftsack.fs.uni-bayreuth.de>
  Infineon Technologies AG
  Munich, Germany

  Jens Lippmann <lippmann@rbg.informatik.tu-darmstadt.de>
  Technische Universit"at Darmstadt.



Changes

# LaTeX2HTML by Nikos Drakos <nikos@cbl.leeds.ac.uk>
# **************************************************
# Change Log ***************************************
# **************************************************
#
# Please use the following style: Add your (unique) initials to the list below.
# Use your initials whereever you make commented changes.
# To get in correlation with the more verbous explanations you make in the
# change log, or if you want to stress meaningful changes, use an additional
# short (about 3 letters) identifier, unique together with your initials,
# eg. jcl-pag. This also helps others to find your changes.
# If any change affects former changes, this *must* be stated in the log,
# together with a reference to the previous log entrie(s).
#
#  nd = Nikos Drakos <nikos@cbl.leeds.ac.uk>
#  ab = Axel Belinfante <Axel.Belinfante@cs.utwente.nl>
#  ac = Andrew Cole <ajcole@cbl.leeds.ac.uk>
#  wb = William Perry <wmperry@edu.indiana.ucs.mango>
#  bh = Bob Hendley <R.J.Hendley@uk.ac.bham.cs>
# rst = Robert S. Thau <rst@edu.mit.ai>
#  dm = David Martland <David.Martland@uk.ac.brunel>
# rjw = Roderick Williams <rodw@cbl.leeds.ac.uk>
#  jr = Jussi Rahola <Jussi.Rahola@csc.fi>
#  ec = Eric Carroll <eric@ca.utoronto.utcc.enfm>
#  tl = <little@com.dec.enet.nuts2u>
#  fz = Franz Vojik <vojik@de.tu-muenchen.informatik>
# spqr= Sebastian Rahtz <spqr@ftp.tex.ac.uk>
#  up = Ulrich Pfeifer <pfeifer@de.uni-dortmund.informatik.ls6>
#  aa = Alberto Accomazzi <alberto@cfa.harvard.edu>
#  wr = Wolfgang Riedel <wolfgang.riedel@de.tu-chemnitz.hrz>
#  alo= Alex Lopez-Ortiz <alopez-o@ca.uwaterloo.barrow>
#  mb = Michael Francis Brantley <brantley@EDU.Princeton>
#  mcb= Michael C. Grant <mcgrant@edu.stanford.rascals>
#  jz = Jelle van Zeijl <jvzeijl@isou17.estec.esa.nl>
#  bs = <boris@cs.rochester.edu>
# jhf = James Freeman <freeman@Cornell-Iowa.edu>
# hpr = Hal Peterson <hrp@com.cray.sdiv>
#  kc = Ken Cox <kenc@com.viewlogic>
#  gp = George Phillips <phillips@ca.ubc.cs>
#  tw = Thomas Wieland <thomas_w@de.uni-bayreuth.mat.btm2x7>
#  dh = Dirk Husemann <Dirk.Husemann@informatik.uni-erlangen.de>
#  bt = Brian  Toonen <toonen@mcs.anl.gov>
#  gg = Graham Gough <graham@cs.man.ac.uk>
#  kz = Kenny Zalewski <zalewk@rpi.edu>
#  tm = Tim MacKenzie (t.mackenzie@trl.oz.au)
#  mk = Martijn Koster (m.koster@nexor.co.uk)
#  ew = Evan Welsh <welsh@uk.ac.ed.epcc>
#  re = Robert Estes <estes@ece.ucdavis.edu>
#  rw = Richard Walker <richard@cs.anu.edu.au>
#  sl = Stefan Lohrum <lohrum@inf.fu-berlin.de>
#  kl = Kris Laporte <kl@sunbim.be>
#  gg = Gordon Greene <greeng3@rpi.edu>
#  pb = Pascal Brisset <brissetp@poly.polytechnique.fr>
#  po = Pieter Olivier <olivierp@fwi.uva.nl>
#  eb = Ethan Bradford <ethanb@ptolemy.astro.washington.edu>
#  ev = Eric Verbeek <wsineric@win.tue.nl>
#  tt = Tom Tromey <tromey@busco.lanl.gov>
#  ms = Mike Stok <Mike.Stok@meiko.concord.ma.us>
#  sk = Stefan Kaes <kaes@lssun001.DresdnerBank.de>
# psm = Peter Magnusson <psm@sics.se>
#  mb = Martin Boyer <gamin@ireq-robot.hydro.qc.ca>
#segal= Mark Segal <segal@spud.asd.sgi.com>
# jmn = Mark Noworolski <jmn@eecs.berkeley.edu>
# meh = Marcus Hennecke <marcush@crc.ricoh.com>
# hws = Herbert W Swan <hswan@perc.Arco.com>
# jkr = Jens Krinke <krinke@ips.cs.tu-bs.de>
# sid = Sidik Isani <isani@cfht.hawaii.edu>
# rml = Rob Malouf <malouf@csli.stanford.edu>
# jtc = Ted Cox <jtcox@gumby.syr.edu>
# pns = Paulo Ney de Souza <desouza@math.berkeley.edu>
# tkm = Tom Miller <tkm@eos.ncsu.edu>
# mg  = Michel Goossens <Michel.Goossens@cern.ch>
# ays = Andreas Schott <schott@rzg.mpg.de>
# jfr = J. Freeman <jfreeman@ms.com>
# mer = Michael Ernst <mernst@cs.rice.edu>
# rrm = Ross Moore <ross@maths.mq.edu.au>
# jcl = Jens Lippmann <lippmann@rbg.informatik.tu-darmstadt.de>
# dtp = D. Taupin <taupin@lps.u-psud.fr>
# mwk = Martin Wilck <martin@kassandra.tropos.de>
# mcm = Barry McMullin <McMullin@EENG.DCU.IE>
# tak = Takashi Tomokiyo <tomokiyo@trc.rwcp.or.jp>
# mro = Marek Rouchal <marek@saftsack.fs.uni-bayreuth.de>
#  bv = Boris Veytsman <boris@plmsc.psu.edu>
#  kr = Keith Refson <Keith.Refson@earth.ox.ac.uk>
#  uw = Uli Wortmann <uli12@bonk.ethz.ch>
# jos = Julius Smith <jos@ccrma.stanford.edu>
#
#
#------------------- Test Suite Manifest ----------------------------------
#
# A listing of current test suites.
# These suites should be run before any new version is released.
# For a detailed description consult the suite's header.
# Feel free to devise further test suites, but *comment* them (see
# jcl-pag.tex header on how to comment).
#
# docs/manual.tex	The l2h manual.
# example/report.tex	An example with segment document parts.
# tests/pages.tex	Tests image generation of l2h.
# tests/rrm-col.tex	See log.
# tests/jcl-pag.tex	See log.
# tests/jcl-env.tex	See log.
# tests/jcl-verb.tex	See log.
#
#--------------------------------------------------------------------------
#
#		CHANGE LOG FOR THE LATEX2HTML RELEASE
#
# What is recorded here are changes to any file of the LaTeX2HTML release
# *meaningful to the user*, such as new options or files, important bug
# fixes etc.
# A more detailed log history can be found within the particular files, near
# the top.
#
#--------- 2013/5
#
#       jos   prefs.pm:
#               - removed -Ppdf option from $DVIPSOPT because it
#                 results in alt-rule.pro being loaded by dvips, which results in
#                 black bars under equations.
#             latex2html.pin:
#               - fixed unequal parens bug in small-caps 
#               - fixed left-aligned address bug (under centered title and author)
#             pstoimg.pin:
#               - fixed grey-background in images bug caused by incorrect search
#                 for "TeXcolorgray" in PS files by pstoimg
#             All changes can be found by searching for "jos:" in source.
#             All of the above bugfixes were reported to latex2html mailing list
#             and various bugzilla posts years ago, some repeatedly, but no action 

config/install.pl
#! /usr/bin/perl -w

###############################################################################
# $Id: install.pl,v 1.12 2002/04/27 11:31:23 RRM Exp $
#
# install.pl
#
# This script is used to install LaTeX2HTML on the user's system
#
# Author: Marek Rouchal <marek@saftsack.fs.uni-bayreuth.de>
#
# This software is part of LaTeX2HTML, originally by Nikos Drakos
# It is published under the GNU Public License and comes without any
# warranty.
#
###############################################################################
# Changes History
#
# $Log: install.pl,v $
# Revision 1.12  2002/04/27 11:31:23  RRM
#      new release version:  2002-1
#
#  --  implements $LATEX2HTMLDIR (shared) and $LATEX2HTMLPLATDIR (unshared)
#      puts unshared l2hconf.pm into $LATEX2HTMLPLATDIR (platform-specific)
#      other modules into $LATEX2HTMLDIR (shareable across platforms)
#
#  --  checks for more graphics conversion utilities, for use with revised
#       graphics.perl  and  graphicx.perl
#
# Revision 1.11  2002/03/07 01:56:19  RRM
#  --  allow the environment variable DESTDIR to be recognise, as a means
#      of giving the installation location; e.g.
#
#         make install DESTDIR=/usr/local/mypackages/
#
#      This is mainly for package-manager software, such as Debian and fink.
#      Thanks to David R Morrison <drm@cgtp.duke.edu>
#      and Daniel Steffen <steffen@maths.mq.edu.au>  for this feature.
#
# Revision 1.10  1999/11/16 00:07:21  MRO
#
# -- minimal change to make $TMP commentable, i.e. unavailable
#
# Revision 1.9  1999/11/09 00:43:42  MRO
#
#
# -- added some changes suggested on the mail list recently: mainly cleanup
#    of -dir option
# -- added installation support for latex2html styles
#
# Revision 1.8  1999/10/26 22:32:57  MRO
#
# -- added a credit
# -- reworked parts of the INSTALL documentation
# -- installation now tries to install styles
#
# Revision 1.7  1999/10/25 21:18:22  MRO
#
# -- added more configure options (Jens' suggestions)
# -- fixed bug in regexp range reported by Achim Haertel
# -- fixed old references in documentation (related to mail list/archive)
#
# Revision 1.6  1999/10/03 18:40:42  MRO
#
# -- some cleanups for beta2
# -- "make check" now checks all Perl code
#
# Revision 1.5  1999/08/30 22:45:09  MRO
#
# -- perl now reports line numbers respective to .pin file - eases
#    code development!
# -- l2hcfg.pm is installed, too for furtjer reference
# -- some minor bugs (hopefully) fixed.
#
# Revision 1.4  1999/06/06 14:24:58  MRO
#
#
# -- many cleanups wrt. to TeXlive
# -- changed $* to /m as far as possible. $* is deprecated in perl5, all
#    occurrences should be removed.
#
# Revision 1.3  1999/06/04 15:30:20  MRO
#
#
# -- fixed errors introduced by cleaning up TMP*
# -- made pstoimg -quiet really quiet
# -- pstoimg -debug now saves intermediate result files
# -- several fixes for OS/2
#
# Revision 1.2  1999/05/19 23:54:07  MRO
#
#
# -- uniquified icons - some of them look a little bit strange, might
#    need to be fixed.
# -- got rid of unlink errors, cleaned up some cosmetics
#
# Revision 1.1  1999/05/11 06:10:03  MRO
#
#
# - merged config stuff, did first tries on Linux. Simple document
#   passes! More test required, have to ger rid of Warnings in texexpand
#
# Revision 1.12  1999/02/23 23:32:35  MRO
#
#
# -- few changes for Win32
#
# Revision 1.11  1999/02/10 01:37:15  MRO
#
#
# -- changed os-dependency structure again - now neat OO modules are
#    used: portable, extensible, neat!
# -- some minor cleanups and bugfixes
#
# Revision 1.10  1999/02/07 23:47:52  MRO
#
#
# -- made install procedure work
#
# Revision 1.9  1998/12/14 22:55:07  MRO
#
#
# -- commit test for MRO
#
# Revision 1.8  1998/12/12 16:39:17  MRO
#
#
# -- cosmetic changes, reworked catching of STDERR in config.pl (Win32)
# -- new configure opt --disable-paths
# -- major cleanups
#
# Revision 1.7  1998/12/07 23:20:00  MRO
#
#
# -- added POD documentation to pstoimg and did a general cleanup
# -- some finetuning of config procedure and modules
#
# Revision 1.6  1998/10/31 14:13:09  MRO
# -- changed OS-dependent module loading strategy: Modules are now located in
#    different (OS-specific) directories nut have the same name: Easier to
#    maintain and cleaner code
# -- Cleaned up config procedure
# -- Extended makefile functionality
#
# Revision 1.5  1998/05/06 22:31:24  latex2html
# -- Enhancements to the config procedure: Added a "generic" target
#    in the Makefile for the TeXlive CD (not perfect yet)
# -- included test for kpsewhich / Web2C
# -- included latest stuff from Override.pm into os_*.pm
#
# Revision 1.4  1998/04/28 22:22:40  latex2html
# - The platform specific stuff is now kept in a separate perl module. This
#   does not introduce significant overhead and enhances maintainability.
#
# Revision 1.3  1998/03/19 23:38:12  latex2html
# -- made pstoimg plug-in compatible with old one (touchwood!)
# -- cleaned up, added some comments
# -- inserted version information output
# -- incorporated patches to make OS/2 run better (thanks Uli)
# -- updated Makefile: make, make test, make install should work now
#
# Revision 1.2  1998/03/17 23:26:47  latex2html
# -- fixed some bugs in config.pl
# -- first version of install.pl finished
#
# Revision 1.1  1998/03/16 23:32:01  latex2html
# -- lots of cosmetic changes and bugfixes, thanks to Uli Wortmann
#    for OS/2 testing
# -- start of install procedure; checks for installation paths implemented
#
###############################################################################

require 5.003;
use strict;
# use diagnostics;
use vars qw(%cfg);

my ($VERSION) = q$Revision: 0.1 $ =~ /:\s*(\S*)/;

# --------------------------------------------------------------------------
# Initialize
# --------------------------------------------------------------------------

use cfgcache;
use L2hos;

my $dd = $cfg{'dd'};

print "install.pl (Revision $VERSION)\n\n";

my $FILECHMOD = 0644; # numbers staring with 0 are octal!
my $DIRCHMOD  = 0755;
my $BINCHMOD  = 0755;

# --------------------------------------------------------------------------
# The distribution's contents
# --------------------------------------------------------------------------

# legal values: bin|lib|plat  store in binary or library directory
#               recurse       recurse for contents (directory only)
#               from=<dir>    use other source directory
#               to=<dir>      use other destination directory
#      separate these flags with ','

my %Install_items = (
  #'Changes'           => 'lib',
  #'FAQ'               => 'lib',
  #'INSTALL'           => 'lib',
  'IndicTeX-HTML'     => 'lib,recurse',
  'L2hos'             => 'lib,recurse',
  'L2hos.pm'          => 'lib',
  #'MANIFEST'          => 'lib',
  #'README'            => 'lib',
  #'README.dvips'      => 'lib',
  #'TODO'            => 'lib',
  'XyMTeX-HTML'       => 'lib,recurse',
  #'config'            => 'lib,recurse',
  #'config.bat'        => 'lib',
  #'configure'         => 'lib',
  #'configure.in'      => 'lib',
  'cweb2html'         => 'lib,recurse',
  'docs'              => 'lib,recurse',
  'dot.latex2html-init' => 'lib',
  'example'           => 'lib,recurse',
  'foilhtml'          => 'lib,recurse',
  # icons are teated specially below
  # the local config is also installed
  'cfgcache.pm'       => 'plat',
  #'latex2html.config' is now l2hconf.pm
  'l2hconf.pm'        => 'plat',
  'makemap'           => 'lib',
  'makeseg'           => 'lib,recurse',
  #'prefs.pm'          => 'lib',
  'readme.hthtml'     => 'lib',
  'styles'            => 'lib,recurse',
  # MRO: Tests are for pre-installation testing
  #'tests'             => 'lib,recurse',
  'texinputs'         => 'lib,recurse',
  #'texlive.pm'        => 'lib',
  'versions'          => 'lib,recurse',
  #'wrapper'           => 'lib,recurse',
);

# --------------------------------------------------------------------------
# Special case: OS dependencies
# --------------------------------------------------------------------------

my @scripts = qw(latex2html pstoimg texexpand);

if($cfg{'texlive'}) {
  #$Install_items{'bin'} = 'lib,recurse';
  # where to install what be built
  my %plat_map = (
        win32 => 'win32:.bat',
        unix => 'i386-linux,rs6000-aix3.2.5,alpha-osf3.2,rs6000-aix4.1,alpha-osf4.0,mips-irix5.3,sparc-solaris2.5.1,hppa1.1-hpux10.10,mips-irix6.2,sparc-sunos4.1.4,powerpc-apple-darwin-current:'
        );
  my $script;
  foreach $script (@scripts) {
    my $plat;
    foreach $plat (qw(unix win32)) {
      my ($d,$ext) = split(/:/, $plat_map{$plat});
      my $ip;
      foreach $ip (split(',',$d)) {
        &install_file("./bin/$plat/$script$ext",
          "$cfg{'PREFIX'}/bin/$ip",$BINCHMOD);
        $Install_items{"bin/$plat/$script$ext"} =
          "to=$cfg{'PREFIX'}/bin/$ip";
      }
    }
    $Install_items{"$script.pl"} = 'lib';
  }
} elsif($cfg{'wrapper'}) {
  foreach(@scripts) {
    $Install_items{"$_$cfg{'exec_extension'}"} = 'bin,from=bin';
    $Install_items{"$_.pl"} = 'lib';
  }
} else {
  foreach(@scripts) {
    $Install_items{"$_$cfg{'exec_extension'}"} = 'bin';
  }
}

# --------------------------------------------------------------------------
# Special case: icons
# --------------------------------------------------------------------------

# icons come in two flavors: GIF and PNG. They go into a icons directory
# in SHLIBDIR (for the -local_icons option) and - if set - to ICONSTORAGE.
# The icons are installed corresponding to the supported image formats.
# If there is no image format supported, GIF icons are copied.

my @icon_types = $cfg{'IMAGE_TYPES'} ? 
  split(/\s+/,$cfg{'IMAGE_TYPES'}) : qw(gif);
my $iconrx = join('|', @icon_types);

my $dest1 = "$cfg{'SHLIBDIR'}${dd}icons";
if((-d $dest1 && !-w _) || (-d $cfg{'SHLIBDIR'} && !-w _)) {
  print STDERR "Error: Cannot install icons in '$dest1': No write permission.\n";
  $dest1 = '';
}
my $dest2 = $cfg{'ICONSTORAGE'} || '';
if(-d $dest2 && !-w $dest2) {
  print STDERR "Error: Cannot install icons in '$dest2': No write permission.\n";
  $dest2 = '';
}
my $dir = "icons";
unless(opendir(DIR,$dir)) {
  print STDERR qq{Error: Could not read directory "$dir": $!\n};
} else {
  my $icon;
  while(defined($icon = readdir(DIR))) {
    next unless($icon =~ /\.($iconrx|html)$/oi);
    my $full = "$dir/$icon";
    next if($icon =~ /^\.{1,2}$/ || !-f $full || !-s _);
    &install_file($full,$dest1,$FILECHMOD,0) if($dest1);
    &install_file($full,$dest2,$FILECHMOD,0) if($dest2);
  }
  closedir(DIR);
}

# --------------------------------------------------------------------------
# Go!
# --------------------------------------------------------------------------

my $item;
foreach $item (sort keys %Install_items) {

  # Special case: pstoimg
  if($item =~ /pstoimg/ && !$cfg{'have_pstoimg'}) {
    print qq{Warning: pstoimg not installed, because build failed.\n};
    next;
  }

  my $from = '.';
  if($Install_items{$item} =~ /(?:^|,)from=([^,]+)(?:,|$)/) {
    $from = $1;
  }

  my $dest = '';
  my $chmod = $FILECHMOD;
  if($Install_items{$item} =~ /(^|,)bin(,|$)/) {
    $dest = $cfg{'BINDIR'};
    $chmod = $BINCHMOD;
  }
  elsif($Install_items{$item} =~ /(^|,)lib(,|$)/) {
    $dest = $cfg{'SHLIBDIR'};
  }
  elsif($Install_items{$item} =~ /(^|,)plat(,|$)/) {
    $dest = $cfg{'LIBDIR'};
  }
  if($Install_items{$item} =~ /(?:^|,)to=([^,]*)(?:,|$)/) {
    $dest = $1;
  }
  
  unless($dest) {
    die qq{Error: No destination for item "$item"\n};
  }
  if($Install_items{$item} =~ /(?:^|,)chmod=(\d+)(,|$)/) {
    $chmod = $1;
  }
  if($Install_items{$item} =~ /recurse/) {
    &install_recurse($item,$dest,$chmod);
  }
  else {
    &install_file("$from/$item",$dest,$chmod);
  }
}

#-----------------------------------------------------------------------------
# try to install LaTeX2HTML style files
#-----------------------------------------------------------------------------

if($cfg{TEXPATH}) {
  print "\nNote: trying to install LaTeX2HTML style files in TeX directory tree\n     ($cfg{TEXPATH})\n";
  unless(mkpath($cfg{TEXPATH})) {
  #my $testpath = $cfg{TEXPATH}; # to strip (latex2)html
  #$testpath =~ s/[$dd$dd][^$dd$dd]*$//;
  #if((-d $cfg{TEXPATH} && !-w _) || (-d $testpath && !-w _)) {
    print STDERR "\nError: Cannot install LaTeX2HTML style files in $cfg{TEXPATH}\n";
  } else {
    my $dir = 'texinputs';
    my $dest = $cfg{TEXPATH};
    unless(opendir(DIR,$dir)) {
      print STDERR qq{Error: Could not read directory "$dir": $!\n};
    } else {
      my $file;
      my $ok = 1;
      while(defined($file = readdir(DIR))) {
        my $full = "$dir/$file";
        next if($file =~ /^\.\.?$/ || !-f $full || !-s _);
        unless(&install_file($full,$cfg{TEXPATH},$FILECHMOD,0)) {
          $ok = 0;
          last;
        }
      }
      closedir(DIR);
      if($ok && $cfg{MKTEXLSR}) {
        print "Info: Running $cfg{MKTEXLSR} to rebuild ls-R database...\n";
        system($cfg{MKTEXLSR});
      }
    }
  }
}
print "Done. Have a lot of fun with LaTeX2HTML!\n";
exit 0;

sub install_file {
  my ($src,$dest,$chmod) = @_;

  # prepend the value of DESTDIR, for package-managers
  $dest = $ENV{'DESTDIR'}.$dest if($ENV{'DESTDIR'});
  my $mkdir = $dest;

  my $file = $src;
  $file =~ s:^.*/::;
  my $full = "$dest$dd$file";

  $src  =~ s:/:$dd:g; # insert correct dir delimiter

  &mkpath($mkdir) || return 0;
  L2hos->Unlink($full); # avoid overwriting the target of symlinks
  unless(L2hos->Copy($src,$full)) {
    return 0;
  } elsif($chmod) {
    unless(chmod($chmod,$full)) {
      print STDERR qq{Warning: chmod of "$full" failed: $!\n};
      return 0;
    }
  }
  print "Info: Installed $full\n";
  1;
}

sub install_recurse {
  my ($dir,$dest,$chmod) = @_;

  my $reldir = $dir;
  $reldir =~ s:^.*/::;

  unless(opendir(DIR,$dir)) {
    print STDERR qq{Error: Could not read directory "$dir": $!\n};
    return 0;
  }
  my @subdirs = ();
  my $item;
  while(defined($item = readdir(DIR))) {
    next if($item =~ /^(\.{1,2}|CVS|\.git)$/);
    my $full = "$dir/$item";
    if(-d $full) {
      push(@subdirs,$full);
    } elsif(-s _) {
      &install_file($full,"$dest$dd$reldir",$chmod);
    } else {
      print "Warning: Skipping $full\n";
    }
  }
  closedir(DIR);
  foreach(@subdirs) {
    &install_recurse($_,"$dest$dd$reldir",$chmod);
  }
  1;
}

sub mkpath {
  my ($dir) = @_;

  unless(-d $dir) {
    if($dir =~ m#^(.*)[$dd$dd]([^$dd$dd]+)$#o) {
      my $parent = $1;
      $parent .= $dd unless($parent =~ /[$dd$dd]/o);
      &mkpath($parent) || return 0; # error
      unless(mkdir($dir,$DIRCHMOD)) {
        print STDERR qq{Error: Could not create directory "$dir": $!\n};
        return 0;
      }
    } else {
      print STDERR qq{Error: Cannot split directory "$dir"\n};
      return 0;
    }
  }
  1;
}

